### PR TITLE
Cps/ Hotfix weak pointers in condition

### PR DESCRIPTION
**Pull request by marcnunezc**
Fixing the potential wall condition after #4835.

I had to remove the "const" definition of the pGetElement() function since the shared_from_this() call was not working. Is that intended (to not work in const functions)?

Thanks!

### CODE REVIEW

**Review by RiccardoRossi — COMMENTED** [review]
i am not sure of what you are trying to do here, nevertheless your fix is not looking good.

maybe you want to check for null?

~~~
KRATOS_ERROR_IF( mpElement.get() == nullptr )
~~~

**Comment by marcnunezc** [comment]
Your suggestion works fine!

**Comment by philbucher** [comment]
I thought one should avoid the shared_from_this?

It somehow is the opposite of what Riccardo did with the intrusive ptr

Maybe the function could return the intrusive ptr?

**Comment by marcnunezc** [comment]
In that case I followed what is done in the FluidDynApp:

https://github.com/KratosMultiphysics/Kratos/blob/5c526642ecb142058ff06acfbe0e878fe3139c39/applications/FluidDynamicsApplication/custom_conditions/fs_generalized_wall_condition.h#L555-L558

(pointing it out in case it needs to be changed there as well)

**Comment by philbucher** [comment]
no worries. This was more a question for @RiccardoRossi (and maybe @jcotela )

Why does this function have to return a shared ptr? Wouldn't a raw ptr be sufficient in this case?

**Comment by RiccardoRossi** [comment]
@philbucher  you are right that if possible we should avoid the shared_from_this.

the good fix woud be to do

~~~
inline GlobalPointer<Element> pGetElement() const
~~~

and then to use it as a global pointer. that would avoid the overhead of updating the counter.

**Comment by marcnunezc** [comment]
> @philbucher you are right that if possible we should avoid the shared_from_this.
> 
> the good fix woud be to do
> 
> ```
> inline GlobalPointer<Element> pGetElement() const
> ```
> 
> and then to use it as a global pointer. that would avoid the overhead of updating the counter.

What would be the syntax to check if its a nullptr in this case?

**Comment by RiccardoRossi** [comment]
identical to the case of a shared pointer

**Comment by marcnunezc** [comment]
When doing this for the `GlobalPointer< Element >`:

`KRATOS_ERROR_IF( mpElement.get() == nullptr )`

I get the followoing error:

```
error: no match for ‘operator==’ (operand types are ‘const Kratos::GlobalPointer<Kratos::Element>’ and ‘std::nullptr_t’)
     KRATOS_ERROR_IF(mpElement == nullptr)

```

**Comment by marcnunezc** [comment]
Ok sorry, you are right, but then again I have to remove the "const" definition in the method for it to work. If I use const I get the following error:
```

error: no matching function for call to ‘Kratos::GlobalPointer<Kratos::Element>::get() const’
     KRATOS_ERROR_IF(mpElement.get() == nullptr)
```

**Comment by philbucher** [comment]
@RiccardoRossi what is the advantage of using a GlobalPointer here? I am not sure if it is better, why not just a raw ptr?

**Comment by RiccardoRossi** [comment]
essentially what i am saying is that you can select a list of ids to be substituted (and for example obtain it from a submodelpart) nevertheless once you made the choice you need to substitute everywherr

**Comment by philbucher** [comment]
> essentially what i am saying is that you can select a list of ids to be substituted (and for example obtain it from a submodelpart) nevertheless once you made the choice you need to substitute everywherr

@RiccardoRossi I think you commented on the wrong PR ...

File: applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.h [hunk 1]
```diff
@@ -292,7 +292,7 @@ class KRATOS_API(COMPRESSIBLE_POTENTIAL_FLOW_APPLICATION) PotentialWallCondition
     ///@name Private Operators
     ///@{
 
-    inline ElementPointerType pGetElement() const;
+    inline GlobalPointer<Element> pGetElement();
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.h]: @RiccardoRossi this is what I was referring to when I had to remove the `const` definition (see cpp).

File: applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.cpp [hunk 2]
```diff
@@ -254,11 +254,11 @@ void PotentialWallCondition<TDim, TNumNodes>::load(Serializer& rSerializer)
 // private operators
 
 template <unsigned int TDim, unsigned int TNumNodes>
-inline Element::Pointer PotentialWallCondition<TDim, TNumNodes>::pGetElement() const
+inline GlobalPointer<Element> PotentialWallCondition<TDim, TNumNodes>::pGetElement()
 {
-    KRATOS_ERROR_IF(!mpElement.lock())
+    KRATOS_ERROR_IF(mpElement.get() == nullptr)
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.cpp:259]: If the pGetElement method is defined as constant, this throws an error in the get() call:

```
error: no matching function for call to ‘Kratos::GlobalPointer<Kratos::Element>::get() const
     KRATOS_ERROR_IF(mpElement.get() == nullptr)
```

**Comment by RiccardoRossi** [comment]
using a raw pointer would be good enough (better actually), only thing is that it would slightly change the semantics (for example it has no get function)

**Comment by RiccardoRossi** [comment]
@marcnunezc Philipp added the const version of get so you don¡'t need any longer to remove the const.

If you can use the raw_pointer as philipp is suggesting

**Comment by marcnunezc** [comment]
Is the raw_pointer the one you get from Element::Pointer?

**Comment by rubenzorrilla** [comment]
> Is the raw_pointer the one you get from Element::Pointer?

Now the one you get from `Element::Pointer` is an intrusive pointer.

**Comment by philbucher** [comment]
you need to dereference the smart-ptr

I don't know by hard right now which is the correct call here unfortunately

The reason why I am suggesting to use it is that you don't need to use a smart pointer (i.e. one that takes care of the memory for you) here, since the element is stored elsewhere (`Mesh`) and there the memory is managed
I.e. you should be save using just a raw ptr

**Comment by marcnunezc** [comment]
Hi guys,

I have tried to use the raw_pointers but it isn't as straightforward as I thought.

Would you mind approving this PR, which uses GlobalPointers, in order to fix the PotentialFlowApp (which is not compiling in the master at the moment)?

I'm using the const pGetElement again and tests are passing

I will open another PR just after this to switch to raw_pointers so I can also ask for help.

**Comment by RiccardoRossi** [comment]
ok on my side. @philbucher i let you approve

File: applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.cpp [hunk 3]
```diff
@@ -254,11 +254,11 @@ void PotentialWallCondition<TDim, TNumNodes>::load(Serializer& rSerializer)
 // private operators
 
 template <unsigned int TDim, unsigned int TNumNodes>
-inline Element::Pointer PotentialWallCondition<TDim, TNumNodes>::pGetElement() const
+inline GlobalPointer<Element> PotentialWallCondition<TDim, TNumNodes>::pGetElement() const
 {
-    KRATOS_ERROR_IF(!mpElement.lock())
+    KRATOS_ERROR_IF(mpElement.get() == nullptr)
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.cpp:259]:
```suggestion
    KRATOS_ERROR_IF(mpElement.get())
```

just as information for you, could be changed in a different PR

**Review by philbucher — APPROVED** [review]
makes sense to me, please go ahead

File: applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.cpp (hunk 3 above)
Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.cpp:259]: You are missing a _NOT or a ! operator right?
```suggestion
    KRATOS_ERROR_IF_NOT(mpElement.get())
```

Comment (philbucher) [reply applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.cpp:259]: yes :)
